loopy/schedule/tree.py

```diff
@@ -164,34 +164,31 @@ def replace_node(self, node: NodeT, new_id: NodeT) -> "Tree[NodeT]":
 
         # {{{ update child to parent
 
-        new_child_to_parent = (self._child_to_parent.delete(node)
-                               .set(new_id, parent))
+        child_to_parent_mut = self._child_to_parent.mutate()
```

Comment:
Because of `mutate`, this now requires the use of `immutables.Map`. Could you just use a normal `dict` for the mutation?

Reply:
Since the internal `immutables.Map`s don't seem to be exposed to the user, maybe it is sufficient to just use `dict` everywhere in this class?

Reply:
With `dict`, $O(N)$ modifications of size $O(1)$ of a tree of size $O(N)$ lead to $O(N^2)$ complexity. `immutables` allows this to remain $O(N)$ (give or take a few $O(\log N)$. So, to my mind, this is the one use case of which I'm aware, where the use of `immutables` is justified (whereas just about everywhere else, it's just a glorified read-only dict, which is total overkill).